src/pressure/VCycleMG.cpp

```diff
@@ -248,171 +215,88 @@ void VCycleMG::pressure(Field &out, Field const &b, real t, bool sync) {
+        for (size_t level = 0; level < m_levels; ++level) {
+            Field *field_residuum0_level = *(m_residuum0 + level);
+            Field *field_error1_level = *(m_error1 + level);
+            Field *field_error1_level_plus_1 = *(m_error1 + level + 1);
+            Field *field_mg_temporal_level = *(m_mg_temporal_solution + level);
+            Field *field_residuum1_level = *(m_residuum1 + level);
+            Field *field_residuum1_level_plus_1 = *(m_residuum1 + level + 1);
+
+            real *data_residuum0_level = field_residuum0_level->data;
+            real *data_residuum1_level_plus_1 = field_residuum1_level_plus_1->data;
+
+            FieldType type_r0 = field_residuum0_level->get_type();
+            FieldType type_r1 = field_residuum1_level_plus_1->get_type();
+
+#pragma acc data present(residuum0_level, error1_level, error1_level_plus_1, \
+                         mg_temporal_level, residuum1_level, residuum1_level_plus_1, \
+                         out)
+            {
                 // smooth
-                Smooth(*f_err1i, *f_mg_tmpi, *f_res1i, i, sync);
+                Smooth(*field_error1_level, *field_mg_temporal_level, *field_residuum1_level, level, sync);
 
                 // calculate residuum
-                Residuum(*f_res0i, *f_err1i, *f_res1i, i, sync);
-                boundary->apply_boundary(d_res0i, i, type_r0, sync);  // for residuum0 only Dirichlet BC
-            }
+                Residuum(*field_residuum0_level, *field_error1_level, *field_residuum1_level, level, sync);
+                boundary->apply_boundary(data_residuum0_level, level, type_r0, sync); // for m_residuum0 only Dirichlet BC
 
-            // restrict
-            Restrict(*f_res1ip, *f_res0i, i, sync);
-            boundary->apply_boundary(d_res1ip, i + 1, type_r1, sync);  // for res only Dirichlet BC
-
-            // set err to zero at next level
-
-            // strides (since GPU needs joined list)
-            // inner start/ end index of level i + 1
-            size_t start_i = boundary->get_inner_list_level_joined_start(i + 1);
-            size_t end_i = boundary->get_inner_list_level_joined_end(i + 1) + 1;
-            // boundary start/ end index of level i + 1
-            size_t start_b = boundary->get_boundary_list_level_joined_start(i + 1);
-            size_t end_b = boundary->get_boundary_list_level_joined_end(i + 1) + 1;
-            // inner
-#pragma acc kernels present(err1ip, d_iList[start_i:(end_i-start_i)]) async
-#pragma acc loop independent
-            for (size_t j = start_i; j < end_i; ++j) {
-                const size_t idx = d_iList[j];
-                d_err1ip[idx] = 0.0;
-            }
+                // restrict
+                Restrict(*field_residuum1_level_plus_1, *field_residuum0_level, level, sync);
+                boundary->apply_boundary(data_residuum1_level_plus_1, level + 1, type_r1, sync); // for res only Dirichlet BC
```

Comment:
two spaces between code and comment are needed.

Reply:
b167247